### PR TITLE
add remote-controllable "pause" state for studies

### PR DESCRIPTION
**Pull request by rhelmer**
Checklist for reviewer:

- [ ] The description should reference a bug or github issue, if relevant.
- [x] There must be a [`CHANGELOG.md`](./CHANGELOG.md) entry for any non-test change.
- [ ] Any change to the NPM commands must be carefully reviewed to make sure it won't break the Add-ons pipeline.
- [ ] Any version increase must follow the [release process](./docs/RELEASE_PROCESS.md).

### CODE REVIEW

**Comment by rhelmer** [comment]
@Dexterp37 there are two commits in this PR - the oldest shows the "disable study add-on with privileged API" approach, and the more recent commit reverts these and sets up message-passing between `rally.js` and the core-addon.

The overall concept is:

1. on add-on startup and when new data arrives, check `studyPaused` field for each study
2. keep the `studyPaused` state in the core add-on and also fire a message to any installed studies
3. do not allow data transmission for "paused" studies (similar to dev mode)

I am imagining study add-ons only need to interact with it like:

```js
rally.addStateListener((state) => { console.debug("fires on pause/resume"));
```

It would be up to the study add-on to actually do the right thing when the callback fires :)

Assuming that all sounds reasonable, I need to sort tests out but wanted to open it up to feedback sooner. I think the problem is that fetching studies happens on every data update now, and the mocks are incomplete.

File: core-addon/Core.js
Comment (Dexterp37) [inline core-addon/Core.js:516]: Can we break this to a separate function?

File: support/rally.js
Comment (Dexterp37) [inline support/rally.js:89]: Mh, ok, what's up with these changes here? Why is your IDE unhappy with our rules? Do we need to change our rules? :)

File: support/rally.js
Comment (Dexterp37) [inline support/rally.js:140]: Same here :)Why is your IDE unhappy with our rules? Do we need to change our rules? :)

File: support/rally.js
Comment (Dexterp37) [inline support/rally.js:186]: We should make this a mandatory parameter of `initialize`, otherwise folks will forget to call it :)

File: tests/core-addon/unit/Core.test.js
Comment (Dexterp37) [inline tests/core-addon/unit/Core.test.js:36]: Most of the changes to this file are unrelated. Even if our linter is wrong, would you mind reverting it and just keep the relevant changes here?

File: tests/support/rally.test.mjs
Comment (Dexterp37) [inline tests/support/rally.test.mjs:95]: Don't forget to remove :)

File: support/rally.js
Comment (Dexterp37) [inline support/rally.js:213]: Do we really need this to be an `error`? This is not really actionable by users nor developers, or am I missing something?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
Thanks for doing this. I like the approach, I have two main observations:

- can we please keep the linting changes out of this PR? It makes review a bit more complicated that it should be. We can tackle them separately (and finally figure out if our linting is weird or your IDE is misbehaving!)
- instead of providing a function to add an observer, can we please require it at initialization?

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:89]: I suspect our eslint isn't strict enough, it seems fine with or without these changes. Not sure why VSCode isn't using the same exact config if that's the case though?

Anyway happy to just remove these :) I filed an issue on the linting stuff.

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:213]: Hm, I would want people who are testing (or normal users) to be able to report it to us. I think `warn` at least is appropriate, but I'm OK with `debug` here .

Comment (rhelmer) [reply support/rally.js:213]: IIRC Firefox shows `debug` by default in the console but Chrome does not.

On a different note, I am starting to think maybe this check should be in the core add-on code and we shouldn't just expect `rally.js` to do the right thing, that feels safer in the face of a misbehaving add-on.

File: core-addon/Core.js
Comment (rhelmer) [reply core-addon/Core.js:515]: I was thinking I could probably shrink this down if I use a more functional style too :) I'll see what I can come up with.

File: core-addon/Core.js
Comment (Dexterp37) [inline core-addon/Core.js:502]: Should we wrap the  `this._sendMessageToStudy` in the try instead of the whole iteration loop? For example, this would allow us to keep on pausing/resumng other studies even if a message is thrown for an individual study.

File: src/web-extension-api.js
Comment (Dexterp37) [inline src/web-extension-api.js:218]: Since we're not changing the UI (other than hiding studies) based on the paused/resumed stated, let's not leak this logic in the UI. Let's do it in [the function to send the updates](https://github.com/mozilla-rally/rally-core-addon/blob/5faf1f355bf25c92189725725d018b0d7c5a69cb/core-addon/Core.js#L567)

File: support/rally.js
Comment (Dexterp37) [reply support/rally.js:213]: Oh yeah, indeed. See what I'm doing in #352 , because I'm doing the same :)

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:186]: > We should make this a mandatory parameter of `initialize`, otherwise folks will forget to call it :)

Hm, do you mean we'd pass the callback function as part of initialize instead of the rally study needing to call `addStateListener()`?

Comment (Dexterp37) [reply support/rally.js:186]: Yes!

File: tests/support/rally.test.mjs
Comment (rhelmer) [reply tests/support/rally.test.mjs:95]: Not needed now that we're doing it as part of initialize, I guess :)

**Comment by rhelmer** [comment]
@Dexterp37 thanks for all the feedback so far! The last(?) big problem I am running into is that this breaks the `Core.js` "dispatches enrollment messages" test (and likely others too), `this._sendMessageToStudy()` is hanging and I haven't figured out the right way to mock it.

I have the rallyjs tests working, but I am holding off on integration tests until you land #392. 

Everything looks OK to me in manual testing and from the rally.js perspective unit tests, so I suspect it's just something I'm not doing quite right with the mocks in the Core.js unit tests which I know you are more familiar with :)

File: support/rally.js
Comment (rhelmer) [inline support/rally.js:214]: Note to self - move this from `rally.js` into `Core.js`, so we do not need to trust the (potentially misbehaving) study extension.

File: tests/support/rally.test.mjs
Comment (Dexterp37) [inline tests/support/rally.test.mjs:88]: Can we also test that this callback gets called and with the proper argument?

File: tests/support/rally.test.mjs
Comment (Dexterp37) [inline tests/support/rally.test.mjs:106]: Can we also test that this callback gets called and with the proper argument?

File: tests/support/rally.test.mjs
Comment (Dexterp37) [inline tests/support/rally.test.mjs:31]: Can we also add a new test here to make sure that we throw if no callback is provided?

File: core-addon/Core.js
Comment (Dexterp37) [inline core-addon/Core.js:502]: Would you kindly add unit tests for `_sendRunState`?

File: core-addon/Core.js
Comment (Dexterp37) [inline core-addon/Core.js:517]: Are you planning on adding an integration test for this? I'm ok with merging this without an integration test as long as we do in depth manual testng :)

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
This looks good overall, I left a few nits. I'd love to have one final look at this after it is rebased on the latest master.

**Comment by Dexterp37** [comment]
@rhelmer note that the test is failing because `_sendStateUpdateToUI` is awaiting on [this._availableStudies](https://github.com/mozilla-rally/rally-core-addon/blob/9fdbcb120e74d1a25487bf0b22672aa20e93c384/core-addon/Core.js#L597) (which is ok). To make the test pass, we can simply mock that function like this `this.core._sendStateUpdateToUI = async () => {};` , right after [this line](https://github.com/mozilla-rally/rally-core-addon/blob/9fdbcb120e74d1a25487bf0b22672aa20e93c384/tests/core-addon/unit/Core.test.js#L172)

File: core-addon/Core.js
Comment (rhelmer) [reply core-addon/Core.js:517]: Yes, honestly I expect the integration test to be fairly simple, but I want to rebase on top of the improvements you made there recently before I start.

File: tests/support/rally.test.mjs
Comment (rhelmer) [reply tests/support/rally.test.mjs:31]: I meant to bring this up, thanks... since we're using positional arguments (as opposed to an options object or parameter desctructuring etc) then it seems like we'd need to enforce all of the options (throw if any argument is not provided or an incorrect type)

**Comment by rhelmer** [comment]
> @rhelmer note that the test is failing because `_sendStateUpdateToUI` is awaiting on [this._availableStudies](https://github.com/mozilla-rally/rally-core-addon/blob/9fdbcb120e74d1a25487bf0b22672aa20e93c384/core-addon/Core.js#L597) (which is ok). To make the test pass, we can simply mock that function like this `this.core._sendStateUpdateToUI = async () => {};` , right after [this line](https://github.com/mozilla-rally/rally-core-addon/blob/9fdbcb120e74d1a25487bf0b22672aa20e93c384/tests/core-addon/unit/Core.test.js#L172)

This worked great, thanks! I went ahead and rebased both `master` and my branch on my github fork, overall works OK but now there are two new test failures for telemetry tests that were modified in the meantime :) I'll work on that and also adding integration tests today.

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:37]: I meant to bring this up, thanks... since we're using positional arguments (as opposed to an options object or parameter desctructuring etc) then it seems like we'd need to enforce all of the options (throw if any argument is not provided or an incorrect type)

Comment (rhelmer) [reply support/rally.js:37]: Well, my previous comment assumes we want to insist that these are required and not optional arguments. If they are optional, then making sure they are the expected type and such is fine and I'll just do that for now.

Comment (rhelmer) [reply support/rally.js:37]: Oh, it turns out that jsdoc has support for callback functions: https://jsdoc.app/tags-callback.html

Comment (rhelmer) [reply support/rally.js:37]: Ugh, it turns out that the jsdoc `@callback` needs to be registered elsewhere in the file. It's probably just fine for auto-generated docs, but since it won't be above the initialize function here I don't think it's useful. I went ahead and documented it in the jsdoc header for initialize, instead.

File: core-addon/Core.js
Comment (rhelmer) [reply core-addon/Core.js:530]: I think this is a good idea but I ran out of time to do that, should we defer to a separate issue? I'd like to think about this a bit more holistically, I feel like we are mixing styles and it'd be helpful to either go more or less consistently functional :)

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
Last batch of small tweaks. I'll happily r+ once the tests pass!

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:124]: My feeling is that checking to see if the study is paused is additional message overhead (the core only knows that the study *should* be paused, we need to actually send a message to the study to determine its state.)

Sending a single message to the study informing it that it should pause makes sense in this context, since only the study has this state. I might be wrong though I'll think about this a bit more :)

Comment (Dexterp37) [reply support/rally.js:124]: I mean just

```suggestion
  _pause() {
    if (this._state !== STATES.PAUSED) {
    this._stateChangeCallback("pause");
    this._state = STATES.PAUSED;
    }
```

just to make sure studies don't pause twice if we make mistakes in the core addon and send the message twice. The rationale is that we control the quality of the code in the core add-on, but we can't guarantee that a double pause won't create problems in the study right now (

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:224]: Hm. Actually I'd like to defer this one, I think we don't have the information we need. I'll send you a message about this as it is a little involved.

File: support/rally.js
Comment (rhelmer) [reply support/rally.js:124]: Oh shoot right for some reason I thought you were pointing at the `Core.js` and not `rally.js` :) of course this sounds good ty.

File: core-addon/Core.js
Comment (rhelmer) [reply core-addon/Core.js:530]: https://github.com/mozilla-rally/rally-core-addon/issues/420

File: core-addon/Core.js
Comment (Dexterp37) [inline core-addon/Core.js:281]:
```suggestion
          throw new Error(`Core._handleExternalMessage - ${sender.id} is paused and may not send data`);
```

Let's try not to use the term telemetry, as this might get confused with Firefox Telemetry.

**Review by Dexterp37 — APPROVED** [review]
r+ with a small nit

**Pull request merged**
